File: app/src/main/kotlin/tech/mmarca/openvitals/features/imports/applehealth/AppleHealthImportModels.kt
package tech.mmarca.openvitals.features.imports.applehealth

import androidx.health.connect.client.records.Record
import androidx.annotation.StringRes
import java.time.Instant
import java.time.ZoneOffset
import kotlin.math.roundToInt
import kotlin.reflect.KClass
import tech.mmarca.openvitals.R

data class AppleHealthImportResult(
    val parsedRecords: Int,
    val parsedWorkouts: Int,
    val parsedCorrelations: Int,
    val parsedActivitySummaries: Int,
    val convertedRecords: Int,
    val importedRecords: Int,
    val duplicateSkippedRecords: Int,
    val notSelectedRecords: Int,
    val unsupportedElements: Int,
    val skippedRecords: Int,
    val failedRecords: Int,
    val workoutRoutesIncomplete: Boolean = false,
    val typeSummaries: List<AppleHealthImportTypeSummary>,
    val diagnostics: List<AppleHealthImportDiagnostic>,
    val shareableReportText: String,
) {
    val unsupportedRecords: Int get() = unsupportedElements
}

data class AppleHealthImportProgress(
    val phase: AppleHealthImportPhase = AppleHealthImportPhase.QUEUED,
    val parsedRecords: Int = 0,
    val parsedWorkouts: Int = 0,
    val parsedCorrelations: Int = 0,
    val parsedActivitySummaries: Int = 0,
    val convertedRecords: Int = 0,
    val importedRecords: Int = 0,
    val duplicateSkippedRecords: Int = 0,
    val notSelectedRecords: Int = 0,
    val unsupportedElements: Int = 0,
    val skippedRecords: Int = 0,
    val failedRecords: Int = 0,
    val expectedSelectedRecords: Int = 0,
) {
    val parsedElements: Int
        get() = parsedRecords + parsedWorkouts + parsedCorrelations + parsedActivitySummaries

    val selectedPreparedRecords: Int
        get() = (convertedRecords - notSelectedRecords).coerceAtLeast(0)

    val percent: Int?
        get() {
            val total = expectedSelectedRecords.takeIf { it > 0 } ?: return null
            if (phase == AppleHealthImportPhase.COMPLETE) return 100
            val selectedProgress = selectedPreparedRecords.coerceAtMost(total)
            val selectedPercent = (selectedProgress.toDouble() / total * SelectedRecordsPercentCeiling).roundToInt()
            val phaseFloor = when (phase) {
                AppleHealthImportPhase.QUEUED,
                AppleHealthImportPhase.PARSING,
                AppleHealthImportPhase.CONVERTING,
                -> 0
                AppleHealthImportPhase.CHECKING_DUPLICATES -> if (selectedProgress >= total) 88 else 0
                AppleHealthImportPhase.WRITING -> if (selectedProgress >= total) 92 else 0
                AppleHealthImportPhase.FINISHING -> 95
                AppleHealthImportPhase.BUILDING_REPORT -> 98
                AppleHealthImportPhase.COMPLETE -> 100
            }
            return maxOf(selectedPercent, phaseFloor).coerceIn(0, 99)
        }
}

private const val SelectedRecordsPercentCeiling = 88

enum class AppleHealthImportPhase {
    QUEUED,
    PARSING,
    CONVERTING,
    CHECKING_DUPLICATES,
    WRITING,
    FINISHING,
    BUILDING_REPORT,
    COMPLETE,
}

@get:StringRes
val AppleHealthImportPhase.labelRes: Int
    get() = when (this) {
        AppleHealthImportPhase.QUEUED -> R.string.settings_apple_health_import_progress_queued
        AppleHealthImportPhase.PARSING -> R.string.settings_apple_health_import_progress_parsing
        AppleHealthImportPhase.CONVERTING -> R.string.settings_apple_health_import_progress_converting
        AppleHealthImportPhase.CHECKING_DUPLICATES -> R.string.settings_apple_health_import_progress_checking_duplicates
        AppleHealthImportPhase.WRITING -> R.string.settings_apple_health_import_progress_writing
        AppleHealthImportPhase.FINISHING -> R.string.settings_apple_health_import_progress_finishing
        AppleHealthImportPhase.BUILDING_REPORT -> R.string.settings_apple_health_import_progress_building_report
        AppleHealthImportPhase.COMPLETE -> R.string.settings_apple_health_import_progress_complete
    }

data class AppleHealthImportTypeSummary(
    val appleType: String,
    val parsed: Int,
    val converted: Int,
    val imported: Int,
    val duplicateSkipped: Int,
    val notSelected: Int,
    val unsupported: Int,
    val skipped: Int,
    val failed: Int,
)

data class AppleHealthImportAnalysisResult(
    val parsedRecords: Int,
    val parsedWorkouts: Int,
    val parsedCorrelations: Int,
    val parsedActivitySummaries: Int,
    val convertedRecords: Int,
    val unsupportedElements: Int,
    val skippedRecords: Int,
    val failedRecords: Int,
    val categorySummaries: List<AppleHealthImportCategorySummary>,
    val typeSummaries: List<AppleHealthImportTypeSummary>,
    val diagnostics: List<AppleHealthImportDiagnostic>,
    val shareableReportText: String,
) {
    val parsedElements: Int
        get() = parsedRecords + parsedWorkouts + parsedCorrelations + parsedActivitySummaries
}

data class AppleHealthImportCategorySummary(
    val category: AppleHealthImportCategory,
    val convertedRecords: Int,
    val routeSessions: Int = 0,
)

data class AppleHealthExportFingerprint(
    val displayName: String?,
    val size: Long?,
) {
    fun isIdentifiable(): Boolean = displayName != null || size != null
}

enum class AppleHealthImportCategory(
    @param:StringRes val titleRes: Int,
    @param:StringRes val descriptionRes: Int,
) {
    WORKOUTS(
        R.string.settings_apple_health_import_category_workouts,
        R.string.settings_apple_health_import_category_workouts_desc,
    ),
    ACTIVITY(
        R.string.settings_apple_health_import_category_activity,
        R.string.settings_apple_health_import_category_activity_desc,
    ),
    HEART(
        R.string.settings_apple_health_import_category_heart,
        R.string.settings_apple_health_import_category_heart_desc,
    ),
    SLEEP(
        R.string.settings_apple_health_import_category_sleep,
        R.string.settings_apple_health_import_category_sleep_desc,
    ),
    BODY(
        R.string.settings_apple_health_import_category_body,
        R.string.settings_apple_health_import_category_body_desc,
    ),
    VITALS(
        R.string.settings_apple_health_import_category_vitals,
        R.string.settings_apple_health_import_category_vitals_desc,
    ),
    NUTRITION(
        R.string.settings_apple_health_import_category_nutrition,
        R.string.settings_apple_health_import_category_nutrition_desc,
    ),
    HYDRATION(
        R.string.settings_apple_health_import_category_hydration,
        R.string.settings_apple_health_import_category_hydration_desc,
    ),
    MINDFULNESS(
        R.string.settings_apple_health_import_category_mindfulness,
        R.string.settings_apple_health_import_category_mindfulness_desc,
    ),
    CYCLE(
        R.string.settings_apple_health_import_category_cycle,
        R.string.settings_apple_health_import_category_cycle_desc,
    ),
}

data class AppleHealthImportDiagnostic(
    val appleType: String,
    val targetType: String?,
    val reasonCode: String,
    val timeRange: String?,
    val unit: String?,
    val value: String?,
    val detail: String,
)

internal data class AppleHealthImportDiagnosticSummary(
    val appleType: String,
    val targetType: String?,
    val reasonCode: String,
    val detail: String,
    val count: Int,
    val exampleTimeRange: String?,
    val exampleUnit: String?,
    val exampleValue: String?,
)

internal data class AppleHealthDiagnosticSummaryKey(
    val appleType: String,
    val targetType: String?,
    val reasonCode: String,
    val detail: String,
)

internal data class MutableAppleHealthImportDiagnosticSummary(
    val appleType: String,
    val targetType: String?,
    val reasonCode: String,
    val detail: String,
    var count: Int,
    val exampleTimeRange: String?,
    val exampleUnit: String?,
    val exampleValue: String?,
) {
    fun toSummary(): AppleHealthImportDiagnosticSummary =
        AppleHealthImportDiagnosticSummary(
            appleType = appleType,
            targetType = targetType,
            reasonCode = reasonCode,
            detail = detail,
            count = count,
            exampleTimeRange = exampleTimeRange,
            exampleUnit = exampleUnit,
            exampleValue = exampleValue,
        )
}

internal fun MutableMap<AppleHealthDiagnosticSummaryKey, MutableAppleHealthImportDiagnosticSummary>.add(
    diagnostic: AppleHealthImportDiagnostic,
) {
    val key = AppleHealthDiagnosticSummaryKey(
        appleType = diagnostic.appleType,
        targetType = diagnostic.targetType,
        reasonCode = diagnostic.reasonCode,
        detail = diagnostic.detail,
    )
    val existing = this[key]
    if (existing != null) {
        existing.count += 1
    } else {
        this[key] = MutableAppleHealthImportDiagnosticSummary(
            appleType = diagnostic.appleType,
            targetType = diagnostic.targetType,
            reasonCode = diagnostic.reasonCode,
            detail = diagnostic.detail,
            count = 1,
            exampleTimeRange = diagnostic.timeRange,
            exampleUnit = diagnostic.unit,
            exampleValue = diagnostic.value,
        )
    }
}

internal data class AppleParsedExport(
    val records: List<AppleRecord>,
    val workouts: List<AppleWorkout>,
    val correlations: List<AppleCorrelation>,
    val parsedRecords: Int,
    val parsedWorkouts: Int,
    val parsedCorrelations: Int,
    val parsedActivitySummaries: Int,
    val parsedTypeCounts: Map<String, Int>,
    /** Raw control characters removed from export.xml because XML 1.0 forbids them as literal text. */
    val sanitizedControlChars: Int = 0,
    /** Bare `&` characters in export.xml that were auto-escaped to `&amp;` because they weren't part of an entity reference. */
    val sanitizedAmpersands: Int = 0,
    /** A truncated workout-route entry was ignored after export.xml had already been read intact. */
    val workoutRouteArchiveFailure: AppleWorkoutRouteArchiveFailure? = null,
)

internal data class AppleWorkoutRouteArchiveFailure(
    val entryName: String,
    val decompressedBytesRead: Long?,
) {
    val detail: String
        get() = buildString {
            append("The ZIP ended unexpectedly while reading ")
            append(entryName)
            if (decompressedBytesRead != null) {
                append(" after ")
                append(decompressedBytesRead)
                append(" decompressed byte(s)")
            }
            append(". Health records were imported from the intact export.xml, but this route and any remaining ZIP entries were unavailable.")
        }
}

internal data class AppleRecord(
    val type: String,
    val sourceName: String?,
    val sourceVersion: String?,
    val device: String?,
    val unit: String?,
    val creationDate: AppleDateTime?,
    val startDate: AppleDateTime?,
    val endDate: AppleDateTime?,
    val rawValue: String?,
    val numericValue: Double?,
    val metadata: Map<String, String>,
    val correlationType: String? = null,
) {
    val valueForReport: String?
        get() = rawValue?.take(80)
}

internal data class AppleWorkout(
    val workoutActivityType: String,
    val sourceName: String?,
    val sourceVersion: String?,
    val device: String?,
    val creationDate: AppleDateTime?,
    val startDate: AppleDateTime?,
    val endDate: AppleDateTime?,
    val duration: Double?,
    val durationUnit: String?,
    val totalDistance: Double?,
    val totalDistanceUnit: String?,
    val totalEnergyBurned: Double?,
    val totalEnergyBurnedUnit: String?,
    val metadata: Map<String, String>,
    val events: List<AppleWorkoutEvent>,
    val routes: List<AppleWorkoutRouteFile> = emptyList(),
    val routeReferences: Int = 0,
    val routeReferencePaths: List<String> = emptyList(),
) {
    val unavailableRoutePaths: List<String>
        get() {
            val availablePaths = routes.mapTo(hashSetOf()) { it.path }
            return routeReferencePaths.filterNot(availablePaths::contains)
        }
}

internal data class AppleWorkoutRouteFile(
    val path: String,
    val points: List<AppleWorkoutRoutePoint>,
)

internal data class AppleWorkoutRoutePoint(
    val latitude: Double,
    val longitude: Double,
    val altitudeMeters: Double?,
    val horizontalAccuracyMeters: Double?,
    val verticalAccuracyMeters: Double?,
)

internal data class AppleWorkoutEvent(
    val type: String?,
    val date: AppleDateTime?,
    val duration: Double?,
    val durationUnit: String?,
)

internal data class AppleCorrelation(
    val type: String,
    val sourceName: String?,
    val sourceVersion: String?,
    val device: String?,
    val creationDate: AppleDateTime?,
    val startDate: AppleDateTime?,
    val endDate: AppleDateTime?,
    val metadata: Map<String, String>,
    val records: List<AppleRecord>,
)

internal data class AppleDateTime(
    val instant: Instant,
    val offset: ZoneOffset?,
) {
    override fun toString(): String = instant.toString()
}

internal data class ConvertedAppleRecord(
    val appleType: String,
    val targetType: String,
    val fingerprint: String,
    val recordType: KClass<out Record>,
    val record: Record,
    val sourceTimeRange: AppleImportTimeRange,
    val unit: String?,
    val value: String?,
)

internal data class AppleImportTimeRange(
    val start: Instant,
    val end: Instant,
) {
    override fun toString(): String =
        if (start == end) start.toString() else "$start..$end"
}

File: lib/features/imports/applehealth/apple_health_import_models.dart
/// Parsed-export and reporting models for the Apple Health importer, ported from
/// the Kotlin `AppleHealthImportModels.kt` (plus the progress/percent logic).
library;

import 'apple_health_import_records.dart';

/// An instant plus its wall-clock offset, mirroring the Kotlin `AppleDateTime`.
class AppleDateTime {
  const AppleDateTime(this.instant, this.offset);

  /// UTC instant.
  final DateTime instant;

  /// Offset from UTC (`java.time.ZoneOffset`), or `null` if unknown.
  final Duration? offset;

  AppleDateTime copyWith({DateTime? instant, Duration? offset}) =>
      AppleDateTime(instant ?? this.instant, offset ?? this.offset);

  @override
  String toString() => instant.toIso8601String();
}

/// A parsed `<Record>` element.
class AppleRecord {
  const AppleRecord({
    required this.type,
    this.sourceName,
    this.sourceVersion,
    this.device,
    this.unit,
    this.creationDate,
    this.startDate,
    this.endDate,
    this.rawValue,
    this.numericValue,
    this.metadata = const {},
    this.correlationType,
  });

  final String type;
  final String? sourceName;
  final String? sourceVersion;
  final String? device;
  final String? unit;
  final AppleDateTime? creationDate;
  final AppleDateTime? startDate;
  final AppleDateTime? endDate;
  final String? rawValue;
  final double? numericValue;
  final Map<String, String> metadata;
  final String? correlationType;

  String? get valueForReport {
    final raw = rawValue;
    if (raw == null) return null;
    return raw.length <= 80 ? raw : raw.substring(0, 80);
  }

  AppleRecord copyWith({String? correlationType}) => AppleRecord(
        type: type,
        sourceName: sourceName,
        sourceVersion: sourceVersion,
        device: device,
        unit: unit,
        creationDate: creationDate,
        startDate: startDate,
        endDate: endDate,
        rawValue: rawValue,
        numericValue: numericValue,
        metadata: metadata,
        correlationType: correlationType ?? this.correlationType,
      );
}

class AppleWorkoutEvent {
  const AppleWorkoutEvent({
    this.type,
    this.date,
    this.duration,
    this.durationUnit,
  });

  final String? type;
  final AppleDateTime? date;
  final double? duration;
  final String? durationUnit;
}

class AppleWorkoutRoutePoint {
  const AppleWorkoutRoutePoint({
    required this.latitude,
    required this.longitude,
    this.altitudeMeters,
    this.horizontalAccuracyMeters,
    this.verticalAccuracyMeters,
  });

  final double latitude;
  final double longitude;
  final double? altitudeMeters;
  final double? horizontalAccuracyMeters;
  final double? verticalAccuracyMeters;
}

class AppleWorkoutRouteFile {
  const AppleWorkoutRouteFile({required this.path, required this.points});

  final String path;
  final List<AppleWorkoutRoutePoint> points;
}

/// A `workout-routes/*.gpx` entry that could not be read because the ZIP ended
/// part-way through it, *after* `export.xml` had already been read intact
/// (Kotlin `AppleWorkoutRouteArchiveFailure`). The health records still import;
/// this route and every later ZIP entry are dropped.
class AppleWorkoutRouteArchiveFailure {
  const AppleWorkoutRouteArchiveFailure({
    required this.entryName,
    this.decompressedBytesRead,
  });

  final String entryName;
  final int? decompressedBytesRead;

  String get detail {
    final buffer = StringBuffer('The ZIP ended unexpectedly while reading ')
      ..write(entryName);
    if (decompressedBytesRead != null) {
      buffer.write(' after $decompressedBytesRead decompressed byte(s)');
    }
    buffer.write(
      '. Health records were imported from the intact export.xml, but this '
      'route and any remaining ZIP entries were unavailable.',
    );
    return buffer.toString();
  }
}

class AppleWorkout {
  const AppleWorkout({
    required this.workoutActivityType,
    this.sourceName,
    this.sourceVersion,
    this.device,
    this.creationDate,
    this.startDate,
    this.endDate,
    this.duration,
    this.durationUnit,
    this.totalDistance,
    this.totalDistanceUnit,
    this.totalEnergyBurned,
    this.totalEnergyBurnedUnit,
    this.metadata = const {},
    this.events = const [],
    this.routes = const [],
    this.routeReferencePaths = const [],
  });

  final String workoutActivityType;
  final String? sourceName;
  final String? sourceVersion;
  final String? device;
  final AppleDateTime? creationDate;
  final AppleDateTime? startDate;
  final AppleDateTime? endDate;
  final double? duration;
  final String? durationUnit;
  final double? totalDistance;
  final String? totalDistanceUnit;
  final double? totalEnergyBurned;
  final String? totalEnergyBurnedUnit;
  final Map<String, String> metadata;
  final List<AppleWorkoutEvent> events;
  final List<AppleWorkoutRouteFile> routes;

  /// Every `<FileReference path=…>` this workout declared, normalized. Kotlin
  /// 1.9.0 carries the *paths* (not just a count) so an unreadable route can be
  /// named in the report for manual recovery.
  final List<String> routeReferencePaths;

  int get routeReferences => routeReferencePaths.length;

  /// Routes the workout declared but whose GPX never made it out of the ZIP.
  List<String> get unavailableRoutePaths {
    final available = routes.map((route) => route.path).toSet();
    return routeReferencePaths
        .where((path) => !available.contains(path))
        .toList();
  }
}

class AppleCorrelation {
  const AppleCorrelation({
    required this.type,
    this.sourceName,
    this.sourceVersion,
    this.device,
    this.creationDate,
    this.startDate,
    this.endDate,
    this.metadata = const {},
    this.records = const [],
  });

  final String type;
  final String? sourceName;
  final String? sourceVersion;
  final String? device;
  final AppleDateTime? creationDate;
  final AppleDateTime? startDate;
  final AppleDateTime? endDate;
  final Map<String, String> metadata;
  final List<AppleRecord> records;
}

/// The result of a full export parse.
class AppleParsedExport {
  const AppleParsedExport({
    required this.records,
    required this.workouts,
    required this.correlations,
    required this.parsedRecords,
    required this.parsedWorkouts,
    required this.parsedCorrelations,
    required this.parsedActivitySummaries,
    required this.parsedTypeCounts,
    this.sanitizedControlChars = 0,
    this.sanitizedAmpersands = 0,
    this.workoutRouteArchiveFailure,
  });

  final List<AppleRecord> records;
  final List<AppleWorkout> workouts;
  final List<AppleCorrelation> correlations;
  final int parsedRecords;
  final int parsedWorkouts;
  final int parsedCorrelations;
  final int parsedActivitySummaries;
  final Map<String, int> parsedTypeCounts;

  /// Raw control characters removed because XML 1.0 forbids them as text.
  final int sanitizedControlChars;

  /// Bare `&` characters auto-escaped to `&amp;`.
  final int sanitizedAmpersands;

  /// Set when a truncated workout-route entry was ignored *after* export.xml had
  /// already been read intact.
  final AppleWorkoutRouteArchiveFailure? workoutRouteArchiveFailure;

  int get parsedElements =>
      parsedRecords + parsedWorkouts + parsedCorrelations + parsedActivitySummaries;

  AppleParsedExport copyWithRouteArchiveFailure(
    AppleWorkoutRouteArchiveFailure? failure,
  ) =>
      AppleParsedExport(
        records: records,
        workouts: workouts,
        correlations: correlations,
        parsedRecords: parsedRecords,
        parsedWorkouts: parsedWorkouts,
        parsedCorrelations: parsedCorrelations,
        parsedActivitySummaries: parsedActivitySummaries,
        parsedTypeCounts: parsedTypeCounts,
        sanitizedControlChars: sanitizedControlChars,
        sanitizedAmpersands: sanitizedAmpersands,
        workoutRouteArchiveFailure: failure ?? workoutRouteArchiveFailure,
      );
}

class AppleImportTimeRange {
  const AppleImportTimeRange(this.start, this.end);

  final DateTime start;
  final DateTime end;

  @override
  String toString() =>
      start == end ? start.toIso8601String() : '${start.toIso8601String()}..${end.toIso8601String()}';
}

/// One converted record ready for import (Kotlin `ConvertedAppleRecord`).
class ConvertedAppleRecord {
  const ConvertedAppleRecord({
    required this.appleType,
    required this.targetType,
    required this.fingerprint,
    required this.record,
    required this.sourceTimeRange,
    this.unit,
    this.value,
  });

  final String appleType;
  final String targetType;

  /// Deterministic import fingerprint used for in-file dedup fallback.
  final String fingerprint;
  final ImportRecord record;
  final AppleImportTimeRange sourceTimeRange;
  final String? unit;
  final String? value;

  /// The record class name used to group duplicate-check queries.
  String get recordType => targetType;

  /// The record's `clientRecordId`, or `null` when blank.
  String? get clientRecordId =>
      record.clientRecordId.isNotEmpty ? record.clientRecordId : null;
}

enum AppleHealthImportPhase {
  queued,
  parsing,
  converting,
  checkingDuplicates,
  writing,
  finishing,
  buildingReport,
  complete,
}

const int _selectedRecordsPercentCeiling = 88;

class AppleHealthImportProgress {
  const AppleHealthImportProgress({
    this.phase = AppleHealthImportPhase.queued,
    this.parsedRecords = 0,
    this.parsedWorkouts = 0,
    this.parsedCorrelations = 0,
    this.parsedActivitySummaries = 0,
    this.convertedRecords = 0,
    this.importedRecords = 0,
    this.duplicateSkippedRecords = 0,
    this.notSelectedRecords = 0,
    this.unsupportedElements = 0,
    this.skippedRecords = 0,
    this.failedRecords = 0,
    this.expectedSelectedRecords = 0,
  });

  final AppleHealthImportPhase phase;
  final int parsedRecords;
  final int parsedWorkouts;
  final int parsedCorrelations;
  final int parsedActivitySummaries;
  final int convertedRecords;
  final int importedRecords;
  final int duplicateSkippedRecords;
  final int notSelectedRecords;
  final int unsupportedElements;
  final int skippedRecords;
  final int failedRecords;
  final int expectedSelectedRecords;

  AppleHealthImportProgress copyWith({
    AppleHealthImportPhase? phase,
    int? expectedSelectedRecords,
  }) =>
      AppleHealthImportProgress(
        phase: phase ?? this.phase,
        parsedRecords: parsedRecords,
        parsedWorkouts: parsedWorkouts,
        parsedCorrelations: parsedCorrelations,
        parsedActivitySummaries: parsedActivitySummaries,
        convertedRecords: convertedRecords,
        importedRecords: importedRecords,
        duplicateSkippedRecords: duplicateSkippedRecords,
        notSelectedRecords: notSelectedRecords,
        unsupportedElements: unsupportedElements,
        skippedRecords: skippedRecords,
        failedRecords: failedRecords,
        expectedSelectedRecords:
            expectedSelectedRecords ?? this.expectedSelectedRecords,
      );

  int get parsedElements =>
      parsedRecords + parsedWorkouts + parsedCorrelations + parsedActivitySummaries;

  int get selectedPreparedRecords {
    final v = convertedRecords - notSelectedRecords;
    return v < 0 ? 0 : v;
  }

  int? get percent {
    final total = expectedSelectedRecords > 0 ? expectedSelectedRecords : null;
    if (total == null) return null;
    if (phase == AppleHealthImportPhase.complete) return 100;
    final selectedProgress =
        selectedPreparedRecords > total ? total : selectedPreparedRecords;
    final selectedPercent =
        (selectedProgress / total * _selectedRecordsPercentCeiling).round();
    final phaseFloor = switch (phase) {
      AppleHealthImportPhase.queued ||
      AppleHealthImportPhase.parsing ||
      AppleHealthImportPhase.converting =>
        0,
      AppleHealthImportPhase.checkingDuplicates =>
        selectedProgress >= total ? 88 : 0,
      AppleHealthImportPhase.writing => selectedProgress >= total ? 92 : 0,
      AppleHealthImportPhase.finishing => 95,
      AppleHealthImportPhase.buildingReport => 98,
      AppleHealthImportPhase.complete => 100,
    };
    final value = selectedPercent > phaseFloor ? selectedPercent : phaseFloor;
    return value.clamp(0, 99);
  }
}

class AppleHealthImportTypeSummary {
  const AppleHealthImportTypeSummary({
    required this.appleType,
    required this.parsed,
    required this.converted,
    required this.imported,
    required this.duplicateSkipped,
    required this.notSelected,
    required this.unsupported,
    required this.skipped,
    required this.failed,
  });

  final String appleType;
  final int parsed;
  final int converted;
  final int imported;
  final int duplicateSkipped;
  final int notSelected;
  final int unsupported;
  final int skipped;
  final int failed;
}

enum AppleHealthImportCategory {
  workouts,
  activity,
  heart,
  sleep,
  body,
  vitals,
  nutrition,
  hydration,
  mindfulness,
  cycle,
}

extension AppleHealthImportCategoryReport on AppleHealthImportCategory {
  String get reportName => switch (this) {
        AppleHealthImportCategory.workouts => 'Workouts and routes',
        AppleHealthImportCategory.activity => 'Activity metrics',
        AppleHealthImportCategory.heart => 'Heart',
        AppleHealthImportCategory.sleep => 'Sleep',
        AppleHealthImportCategory.body => 'Body measurements',
        AppleHealthImportCategory.vitals => 'Vitals',
        AppleHealthImportCategory.nutrition => 'Nutrition',
        AppleHealthImportCategory.hydration => 'Hydration',
        AppleHealthImportCategory.mindfulness => 'Mindfulness',
        AppleHealthImportCategory.cycle => 'Cycle tracking',
      };
}

const Set<AppleHealthImportCategory> allAppleHealthImportCategories = {
  AppleHealthImportCategory.workouts,
  AppleHealthImportCategory.activity,
  AppleHealthImportCategory.heart,
  AppleHealthImportCategory.sleep,
  AppleHealthImportCategory.body,
  AppleHealthImportCategory.vitals,
  AppleHealthImportCategory.nutrition,
  AppleHealthImportCategory.hydration,
  AppleHealthImportCategory.mindfulness,
  AppleHealthImportCategory.cycle,
};

class AppleHealthImportCategorySummary {
  const AppleHealthImportCategorySummary({
    required this.category,
    required this.convertedRecords,
    this.routeSessions = 0,
  });

  final AppleHealthImportCategory category;
  final int convertedRecords;
  final int routeSessions;
}

class AppleHealthImportDiagnostic {
  const AppleHealthImportDiagnostic({
    required this.appleType,
    required this.targetType,
    required this.reasonCode,
    required this.timeRange,
    required this.unit,
    required this.value,
    required this.detail,
  });

  final String appleType;
  final String? targetType;
  final String reasonCode;
  final String? timeRange;
  final String? unit;
  final String? value;
  final String detail;
}

class AppleHealthImportDiagnosticSummary {
  AppleHealthImportDiagnosticSummary({
    required this.appleType,
    required this.targetType,
    required this.reasonCode,
    required this.detail,
    required this.count,
    required this.exampleTimeRange,
    required this.exampleUnit,
    required this.exampleValue,
  });

  final String appleType;
  final String? targetType;
  final String reasonCode;
  final String detail;
  int count;
  final String? exampleTimeRange;
  final String? exampleUnit;
  final String? exampleValue;
}

class AppleHealthImportResult {
  const AppleHealthImportResult({
    required this.parsedRecords,
    required this.parsedWorkouts,
    required this.parsedCorrelations,
    required this.parsedActivitySummaries,
    required this.convertedRecords,
    required this.importedRecords,
    required this.duplicateSkippedRecords,
    required this.notSelectedRecords,
    required this.unsupportedElements,
    required this.skippedRecords,
    required this.failedRecords,
    required this.typeSummaries,
    required this.diagnostics,
    required this.shareableReportText,
    this.workoutRoutesIncomplete = false,
  });

  final int parsedRecords;
  final int parsedWorkouts;
  final int parsedCorrelations;
  final int parsedActivitySummaries;
  final int convertedRecords;
  final int importedRecords;
  final int duplicateSkippedRecords;
  final int notSelectedRecords;
  final int unsupportedElements;
  final int skippedRecords;
  final int failedRecords;

  /// A workout-route entry was unreadable because the ZIP ended unexpectedly;
  /// health records still imported (Kotlin `workoutRoutesIncomplete`).
  final bool workoutRoutesIncomplete;
  final List<AppleHealthImportTypeSummary> typeSummaries;
  final List<AppleHealthImportDiagnostic> diagnostics;
  final String shareableReportText;

  int get unsupportedRecords => unsupportedElements;
}

class AppleHealthImportAnalysisResult {
  const AppleHealthImportAnalysisResult({
    required this.parsedRecords,
    required this.parsedWorkouts,
    required this.parsedCorrelations,
    required this.parsedActivitySummaries,
    required this.convertedRecords,
    required this.unsupportedElements,
    required this.skippedRecords,
    required this.failedRecords,
    required this.categorySummaries,
    required this.typeSummaries,
    required this.diagnostics,
    required this.shareableReportText,
  });

  final int parsedRecords;
  final int parsedWorkouts;
  final int parsedCorrelations;
  final int parsedActivitySummaries;
  final int convertedRecords;
  final int unsupportedElements;
  final int skippedRecords;
  final int failedRecords;
  final List<AppleHealthImportCategorySummary> categorySummaries;
  final List<AppleHealthImportTypeSummary> typeSummaries;
  final List<AppleHealthImportDiagnostic> diagnostics;
  final String shareableReportText;

  int get parsedElements =>
      parsedRecords + parsedWorkouts + parsedCorrelations + parsedActivitySummaries;
}

/// Mutable per-Apple-type accounting used during conversion/import (Kotlin
/// `MutableAppleImportTypeStats`).
class MutableAppleImportTypeStats {
  int parsed = 0;
  int converted = 0;
  int imported = 0;
  int duplicateSkipped = 0;
  int notSelected = 0;
  int unsupported = 0;
  int skipped = 0;
  int failed = 0;
}
